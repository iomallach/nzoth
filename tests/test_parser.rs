mod helpers;
use std::cell::RefCell;

use helpers::SnapshotErrors;
use insta;
use nzoth::{parser::Parser, source::SourceFile};

#[test]
fn test_valid_let_declarations() {
    insta::glob!("resources/parser_cases/valid_let_declarations.nz", |path| {
        let contents = std::fs::read_to_string(path).expect("Expect path exists");
        let source = RefCell::new(SourceFile::new(
            0,
            path.file_name()
                .expect("Expect a file")
                .to_str()
                .expect("No fail")
                .to_string(),
            contents.as_str(),
        ));

        let mut parser = Parser::new(&source);
        let program = parser.parse();
        assert_eq!(2, program.nodes.len());

        let output = format!("{:#?}", program);

        insta::assert_snapshot!(output);
    })
}

#[test]
fn test_valid_identifier() {
    insta::glob!("resources/parser_cases/valid_identifiers.nz", |path| {
        let contents = std::fs::read_to_string(path).expect("Expect path exists");
        let source = RefCell::new(SourceFile::new(
            0,
            path.file_name()
                .expect("Expect a file")
                .to_str()
                .expect("No fail")
                .to_string(),
            contents.as_str(),
        ));

        let mut parser = Parser::new(&source);
        let program = parser.parse();

        let output = format!("{:#?}", program);
        insta::assert_snapshot!(output);
    })
}

#[test]
fn test_valid_assignments() {
    insta::glob!("resources/parser_cases/valid_assignments.nz", |path| {
        let contents = std::fs::read_to_string(path).expect("Expect path exists");
        let source = RefCell::new(SourceFile::new(
            0,
            path.file_name()
                .expect("Expect a file")
                .to_str()
                .expect("No fail")
                .to_string(),
            contents.as_str(),
        ));

        let mut parser = Parser::new(&source);
        let program = parser.parse();

        let output = format!("{:#?}", program);
        insta::assert_snapshot!(output);
    })
}

#[test]
fn test_invalid_assignments() {
    insta::glob!("resources/parser_cases/invalid_assignments.nz", |path| {
        let contents = std::fs::read_to_string(path).expect("Expect path exists");
        let source = RefCell::new(SourceFile::new(
            0,
            path.file_name()
                .expect("Expect a file")
                .to_str()
                .expect("No fail")
                .to_string(),
            contents.as_str(),
        ));

        let mut parser = Parser::new(&source);
        _ = parser.parse();

        insta::assert_snapshot!(SnapshotErrors(parser.errors()));
    })
}
