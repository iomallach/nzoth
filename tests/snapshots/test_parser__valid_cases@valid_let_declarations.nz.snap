---
source: tests/test_parser.rs
input_file: tests/resources/parser_cases/valid_let_declarations.nz
---
Program {
    nodes: [
        Statement(
            LetDeclaration(
                LetDeclaration {
                    identifier: "foo",
                    expression: NumericLiteral(
                        Integer(
                            1,
                        ),
                        Span {
                            start: Position {
                                offset: 10,
                                line: 0,
                                column: 10,
                            },
                            end: Position {
                                offset: 11,
                                line: 0,
                                column: 11,
                            },
                            file: FileId(
                                0,
                            ),
                        },
                    ),
                    span: Span {
                        start: Position {
                            offset: 0,
                            line: 0,
                            column: 0,
                        },
                        end: Position {
                            offset: 11,
                            line: 0,
                            column: 11,
                        },
                        file: FileId(
                            0,
                        ),
                    },
                    ty: None,
                },
            ),
        ),
        Statement(
            LetDeclaration(
                LetDeclaration {
                    identifier: "foo",
                    expression: NumericLiteral(
                        Integer(
                            1,
                        ),
                        Span {
                            start: Position {
                                offset: 30,
                                line: 1,
                                column: 17,
                            },
                            end: Position {
                                offset: 31,
                                line: 1,
                                column: 18,
                            },
                            file: FileId(
                                0,
                            ),
                        },
                    ),
                    span: Span {
                        start: Position {
                            offset: 13,
                            line: 1,
                            column: 0,
                        },
                        end: Position {
                            offset: 31,
                            line: 1,
                            column: 18,
                        },
                        file: FileId(
                            0,
                        ),
                    },
                    ty: Some(
                        Name(
                            "int",
                        ),
                    ),
                },
            ),
        ),
        Statement(
            LetDeclaration(
                LetDeclaration {
                    identifier: "bar",
                    expression: FunctionCall(
                        Identifier(
                            "foo",
                            Span {
                                start: Position {
                                    offset: 43,
                                    line: 2,
                                    column: 10,
                                },
                                end: Position {
                                    offset: 46,
                                    line: 2,
                                    column: 13,
                                },
                                file: FileId(
                                    0,
                                ),
                            },
                        ),
                        [
                            NumericLiteral(
                                Float(
                                    12.3,
                                ),
                                Span {
                                    start: Position {
                                        offset: 47,
                                        line: 2,
                                        column: 14,
                                    },
                                    end: Position {
                                        offset: 51,
                                        line: 2,
                                        column: 18,
                                    },
                                    file: FileId(
                                        0,
                                    ),
                                },
                            ),
                            Infix(
                                Identifier(
                                    "q",
                                    Span {
                                        start: Position {
                                            offset: 53,
                                            line: 2,
                                            column: 20,
                                        },
                                        end: Position {
                                            offset: 54,
                                            line: 2,
                                            column: 21,
                                        },
                                        file: FileId(
                                            0,
                                        ),
                                    },
                                ),
                                Add,
                                NumericLiteral(
                                    Integer(
                                        4,
                                    ),
                                    Span {
                                        start: Position {
                                            offset: 57,
                                            line: 2,
                                            column: 24,
                                        },
                                        end: Position {
                                            offset: 58,
                                            line: 2,
                                            column: 25,
                                        },
                                        file: FileId(
                                            0,
                                        ),
                                    },
                                ),
                                Span {
                                    start: Position {
                                        offset: 53,
                                        line: 2,
                                        column: 20,
                                    },
                                    end: Position {
                                        offset: 54,
                                        line: 2,
                                        column: 21,
                                    },
                                    file: FileId(
                                        0,
                                    ),
                                },
                            ),
                        ],
                        Span {
                            start: Position {
                                offset: 43,
                                line: 2,
                                column: 10,
                            },
                            end: Position {
                                offset: 46,
                                line: 2,
                                column: 13,
                            },
                            file: FileId(
                                0,
                            ),
                        },
                    ),
                    span: Span {
                        start: Position {
                            offset: 33,
                            line: 2,
                            column: 0,
                        },
                        end: Position {
                            offset: 46,
                            line: 2,
                            column: 13,
                        },
                        file: FileId(
                            0,
                        ),
                    },
                    ty: None,
                },
            ),
        ),
    ],
}
